docs_src/design/adr/security/0027-authentication.md

```diff
@@ -0,0 +1,257 @@
+to delegate confirmation of JWT validity to Vault,
+as a mitigation for microservice-level implementation errors.
+TLS termination at the reverse proxy shall be enabled by default
+so as to be consistent with
+[ADR 0015 - Encryption between microservices](./0015-in-cluster-tls.md).
+
+Behind the proxy, there are two major changes:
+
+1. Every EdgeX service, when security is enabled,
+   requires a JWT be passed as part of the HTTP request
+   that is validated using Vault's token introspection endpoint,
+   or manually validated based on published signature keys.
+
+2. Every EdgeX service, when security is enabled,
+   uses a Vault-supplied JWT to authenticate
+   outgoing calls to peer EdgeX services.
+   The original caller's identity may be passed through
+   at the developers' discretion for microservice chaining scenarios.
+
+The new TO-BE architecture is diagrammed in the following figure:
+
+![TO-BE Architecture](0027-to-be.jpg)
+
+
+### Implementation pre-requisites
+
+This ADR assumes a minor refactoring to the security bootstrapping
+components use the Vault identity API and one or more authentication engines
+to issue identity-based Vault tokens instead of raw Vault tokens.
+Affected services include,
+`go-mod-secrets` (configure identity, issue and validate JWT's),
+`security-secretstore-setup`,
+`security-file-token-provider`,
+and `security-spiffe-token-provider`.
+
+This refactoring results in several benefits:
+
+* It de-privileges `security-secretstore-setup`'s use of Vault,
+  which currently requires Vault "sudo" capability to issue raw Vault tokens.
+  (This is a blocking issue for customers that want to bring their own Vault.)
+
+* An external user identity could be authenticated by
+  an external service, such as [Auth0](https://auth0.com).
+  Alternatively, username/password or AppRole authentication
+  could be used if an external source of identity is not available.
+  This is viewed as beneficial, as downstream EdgeX deployments
+  are already building their own similar integrations.
+
+* An internal service identity could be authenticated by
+  a Kubernetes service account token.  This could eliminate
+  the requirement to pre-distribute Vault tokens to services
+  via a shared filesystem volume, simplifying Kubernetes-based
+  deployments of EdgeX.
+
+* As an added bonus, Vault supports longer JWT key sizes than the Kong JWT plugin.
+
+Additionally, `security-bootstrapper` will need to modified to not
+block on availability of Postgres before issuing the ready-to-run signal.
+(This change is already completed.)
+
+
+### High-level list of changes
+
+The following list of changes is derived from the proof of concept implementation
+to actually effect the change (besides the prerequisite changes above):
+
+- Kong and Postgres is removed from compose files and snaps.
+
+- Add an NGINX reverse proxy with using the proxy auth module.
+
+- Create a new `security-proxyauth` service to check the incoming JWT for validity.
```

Comment:
the service name should be `security-proxy-auth` for the consistency

Reply:
Updated.